### DISCUSSION TITLE
How to implement a rate limiter

### DISCUSSION THREAD

**Discussion by FleetAdmiralJakob** [trpc/trpc#4672] [Q&A] [answered]
Hey, how can I implement a rate limiter into my API to protect it?

**Comment by prenaissance** [comment, marked as answer]
The easiest way would be to do that with a reverse proxy, like [Nginx](https://www.nginx.com/blog/rate-limiting-nginx/).
To do that with TRPC specifically, you could use middleware. Here's an example, with an ioredis instance in the context:
```ts
const t = initTRPC.context<typeof createTRPCContext>().create({
  transformer: superjson,
  errorFormatter({ shape, error }) {
    return {
      ...shape,
      data: {
        ...shape.data,
        zodError:
          error.cause instanceof ZodError ? error.cause.flatten() : null,
      },
    };
  },
});

const limit = 10;
const window_seconds = 60;

const rateLimit = t.middleware(async ({ ctx, next }) => {
  const { redis } = ctx;
  const interval = Math.floor(Date.now() / 1000 / window_seconds);
  const key = `rate-limit:${ctx.session.user.id}:${interval}`;

  const current = await redis.incr(key);
  await redis.expire(key, window_seconds, "NX");

  if (current > limit) {
    throw new TRPCError({
      code: "TOO_MANY_REQUESTS",
      message: "You are being rate limited.",
    });
  }

  return next();
});

export const rateLimitedProcedure = t.procedure.use(rateLimit);
```

**Reply by FleetAdmiralJakob** [reply]
Where can I do this?
I don't have any experiences with middlewares

**Reply by prenaissance** [reply]
You'd use this for the procedures you'd want to rate limit and the middleware will run before every such procedure.
[https://trpc.io/docs/server/middlewares](https://trpc.io/docs/server/middlewares)

**Reply by FleetAdmiralJakob** [reply]
Thank you very much for help!!! I think I figured the middleware stuff out. But I got type errors in your code: 
![image](https://github.com/trpc/trpc/assets/77451351/198b90a9-f256-4bb4-bba6-072857243bb5)

**Reply by prenaissance** [reply]
No problem. That's an example, assuming auth.js is in the project to inject the session in context. You may as well use the IP address, a certain cookie or another form of identification to use in the key.

**Reply by FleetAdmiralJakob** [reply]
Hey @prenaissance, maybe I'm stupid but how does this ctx thingy work? Like I tried this whole hour do get an upstash rate limiter to work but I think I have to grab the IP Adress out of the ctx but I don't know how this works.
Can you say to me how I can grab data out of ctx?

**Reply by FleetAdmiralJakob** [reply]
Thank you @prenaissance I figured it out.
